pkg/gadgets/trace/dns/tracer/bpf/dns.c

```diff
@@ -204,6 +232,43 @@ output_dns_event(struct __sk_buff *skb, union dnsflags flags, __u32 name_len, __
 	int anaddrcount = load_addresses(skb, ancount, anoffset, event);
 	event->anaddrcount = anaddrcount;
 
+	// Calculate latency:
+	//
+	// Filter by mount_ns_id and pkt_type to track the latency from when a query is sent from a container
+	// and when a response to the query is received by that same container.
+	//
+	// * On DNS query sent from a container namespace (mount_ns_id != 0, qr == DNS_QR_QUERY, and pkt_type == OUTGOING),
+	//   store the query timestamp in a map.
+	//
+	// * On DNS response received in the same container namespace (mount_ns_id != 0, qr == DNS_QR_RESP, and pkt_type == HOST)
+	//   retrieve/delete the query timestamp and set the latency field on the event.
+	//
+	// A garbage collection thread running in userspace periodically scans for keys with old timestamps
+	// to free space occupied by queries that never receive a response.
+	//
+	// Skip this if skb_val == NULL (gadget_socket_lookup did not set pid_tgid we use in the query key)
+	// or if event->timestamp == 0 (kernels before 5.8 don't support bpf_ktime_get_boot_ns, and the patched
+	// version IG injects always returns zero).
+	if (skb_val != NULL && event->timestamp > 0) {
+		struct query_key_t query_key = {
+			.pid_tgid = skb_val->pid_tgid,
```

Comment:
What do you think about removing the `skb_val != NULL` check and just use 0 in case it isn't available?

```suggestion
	if (event->timestamp > 0) {
		struct query_key_t query_key = {
			.pid_tgid = skb_val ? skb_val->pid_tgid : 0,
```

IMO that's a good fallback when the socket enricher it's not available. @alban  do you have comments on this?

Reply:
Would that get confused with `trace dns -A`? I'm thinking about what happens to a DNS packet that goes from client -> coredns -> upstream DNS and back again. I think the DNS ID is the same when coredns forwards the query upstream, and if we're setting pid_tgid to zero then query_key would also be the same.

If all the containers are using the same query_key, then I think latency would be reported for the response back to coredns, but calculated from when the request was sent from the client (not when it was sent from coredns)